### ISSUE TITLE
It is possible to send chunked response from fastify service?

### ISSUE THREAD

**Issue by mihao2002** [fastify/fastify#805]
I am trying to see if it is possible to send chunked response (Transfer-Encoding : chunked) from a fastify service.

I read the fastify code, and it seems to me that the only way to send response is to call "reply.send(payload)", but when it is called, the service response will be ended, leaving no chance to append another chunk to the response stream.

Please advise.

Thanks

**Comment by mcollina** [comment]
Currently it is not. You cannot do subsequent calls to .send() during a single route/hook. I think you can achieve what you need by passing a Readable stream to send.

Can you add some code that uses either node core or express so that I understand what you want to achieve?

**Comment by mihao2002** [comment]
Our web service needs to send large volume of data back to client, and we want to pipe portion of data to client whenever it is ready. So ideally, what I want to achieve can be illustrated by the following pseudo-code:

onPost = (req,res)=>
{
 ...
    while (!data.End)
    {
        var readyPortion = data.GetDataReadyAndNotProcessed();

       // This one will send the portion of data back to client with "Transfer-Encoding : chunked", 
       // so client can start to process it
       res.sendPartial(readyPortion)
    }

    // When there is no further data, mark the end of the response
    res.Send();
}

You mentioned that I can use readable stream. Just to clarify, do you mean, if I feed the .send method a readable stream, the .send method will loop on the stream, pipe whatever data in the stream to client till the stream reaches its end. While on there other hand, I can use a separate thread to fill that stream to achieve the parallelism what I want. (But I feel like that the .send method needs to load the entire stream before sending any bit to client, because it has to calculate the content length header, hasn't it? Sorry, just my guessing here, I haven't read though the code yet.)

**Comment by jsumners** [comment]
Sending a stream is the way to do what you want. HTTP is a stream protocol, after all.

**Comment by mihao2002** [comment]
Then how can I deal with the missing content-length header in the response? Meanwhile, without the "transfer-encoding: chunked" support, client has no idea on the boundary of each chunk.

**Comment by nwoltman** [comment]
When sending a stream, the `Transfer-Encoding: chunked` header is automatically set.

If Fastify's [`reply.send()`](https://github.com/fastify/fastify/blob/master/docs/Reply.md#send) method doesn't suit your needs though, you can always use Node's native [`res`](https://nodejs.org/dist/latest-v8.x/docs/api/http.html#http_class_http_serverresponse) object which is available on [`reply`](https://github.com/fastify/fastify/blob/master/docs/Reply.md) as `reply.res`.

**Comment by mcollina** [comment]
Can you please add some javascript code on how you are doing this right now without Fastify?

**Comment by mihao2002** [comment]
Here is it. I verified that node core can support chunked http response. Bellow is the sample server code:

var http = require('http');
var stream = require('stream');

http.createServer((req, res) => {
	res.writeHead(200, { 'Content-Type': 'text/html' });

	// Create a buffer to hold the response chunks
	var buffer = new stream.Readable();
	buffer._read = ()=>{};

	// Pipe the buffer to the response stream.
	buffer.pipe(res);

	// Generate 5 chunks with 1 second interval
	var count = 5;
	var emit = () => {
		var data = `Hello World ${count}`;
		console.log(`sending "${data}"`);
		buffer.push(data);

		count--;
		if (count > 0) {
			setTimeout(emit, 1000);
		}
		else {
			console.log('end sending.');
			buffer.push(null);
		}
	};

	emit();
}).listen(8080);

**Comment by mcollina** [comment]
This is working as expected:

```js

var Fastify = require('fastify');
var stream = require('stream');
var fastify = Fastify()

fastify.get('/', function (request, reply) {

	// Create a buffer to hold the response chunks
	var buffer = new stream.Readable();
	buffer._read = ()=>{};

	// Generate 5 chunks with 1 second interval
	var count = 5;
	var emit = () => {
		var data = `Hello World ${count}`;
		console.log(`sending "${data}"`);
		buffer.push(data);

		count--;
		if (count > 0) {
			setTimeout(emit, 1000);
		}
		else {
			console.log('end sending.');
			buffer.push(null);
		}
	};

	emit();
	reply.type('text/html').send(buffer)
})

fastify.listen(8080);
```

**Comment by fas3r** [comment]
Hello @mcollina , 

Following your example, would it be possible to `stream` the request as a buffer without rewriting the request ( header, body, qs, ..) ? Similar to what you do in `fastify-reply-from`  before forwarding the request to the backend server .

I'm wondering if it could be possible to stream the request in a Duplex stream or a Transform using `readable-stream` to add what need to be added in the header (if anything) and send that over a tcp/tls connection which is already established to a backend server.

It could avoid the rewriting part.

Thanks.

**Comment by AlexLeung** [comment]
People need to stop recommending that you build an in-memory buffer by appending chunks, then sending the total concatenated string. If you're streaming a 10GB file, this strategy will try and add the entire 10GB into memory, which completely defeats the purpose of streaming.

Instead just pipe `reply.raw`.

**Comment by Eomm** [comment]
This issue is quite old, and the suggestion was to use the stream (that will not load all the data into memory).

Right now you don't need to manage the raw response since you may write 

```
reply.send(require('fs').createReadStream('./10GB-File.txt'))
```

and this will stream the file without filling the memory.

**Comment by JoshuaRamsamooj** [comment]
I was able to solve a similar problem by setting the `filename` parameter. Maybe FormData is not able to deduce the filename from a passthrough stream?

``` 
var data = new FormData();
data.append('file', passThroughStream, {
  filename: 'some-filename'
});
```

**Comment by Yakumwamba** [comment]
Is anybody having an issue with this in 2022?

**Comment by jcbbb** [comment]
@Yakumwamba Yes, I cannot get it to work on safari, and it doesn't work on Firefox if you don't add `charset=utf-8` with `Content-Type`. Chrome works just fine.

**Comment by woss** [comment]
https://www.fastify.io/docs/latest/Reference/Reply/#streams << 'nuf said :)
